language.go

```diff
@@ -116,6 +137,54 @@ func PrefixExtension(r rune, ext func(context.Context, *Parser) (Evaluable, erro
 	return l
 }
 
+// PrefixLanguage switches the parsing language from the current language to the
+// given language. When the subordinate language cannot consume any more input
+// from the parser, control is returned to the current language and the callback
+// is invoked to decide how to continue.
+func PrefixLanguage(r rune, sub Language, ext func(context.Context, *Parser, Evaluable) (Evaluable, error)) Language {
+	l := newLanguage()
+	l.prefixes[r] = func(c context.Context, p *Parser) (Evaluable, error) {
+		eval, err := sub.ChainEvaluable(c, p)
+		if err != nil {
+			return nil, err
+		}
+
+		return ext(c, p, eval)
+	}
+	return l
+}
+
+// Init is a language that does no parsing, but invokes the given function when
+// parsing starts. It is incumbent upon the function to call ParseExpression to
+// continue parsing.
+//
+// This function can be used to customize the parser settings, such as
+// whitespace or ident behavior.
+func Init(ext func(context.Context, *Parser) (Evaluable, error)) Language {
+	l := newLanguage()
+	l.init = ext
+	return l
+}
+
+// Alternate is a language that runs the given function if no other prefix
+// matches.
+func Alternate(ext func(context.Context, *Parser) (Evaluable, error)) Language {
+	l := newLanguage()
+	l.alternate = ext
+	return l
+}
+
+// Late allows late-binding of a language so that languages can be
+// self-referencing. The return value of the passed function will be bound to
+// the language.
+func Late(fn func(late Language) Language) (binding Language) {
```

Comment:
I'm pretty sure you don't need this and could use somethig like

```go
var a = PrefixExtension('x', parseX)
var b = PrefixExtension('y', parseY)

func parseX func(context.Context, *Parser) (Evaluable, error){
  p.ParseSublanguage(b)
}

func parseY func(context.Context, *Parser) (Evaluable, error){
  p.ParseSublanguage(a)
}
```

Reply:
Hmm, indeed, both this function and `PrefixLanguage` are trivially built using the other additions, and with entirely public interfaces.

Removing `PrefixLanguage` seems harmless, but my only concern about removing `Late` is that users may not discover such a mechanism for implementing recursive invocation by themselves. In many parser combinator libraries, in addition to the base set of parser constructs, they also expose some "patterns" that are built on top of the primitives, like `Sequence`, `Many`, `Optional`, etc. Do you think it would it make sense to have a `patterns.go` that has these type of utility language constructors in it?

For now, I've removed it altogether, but let me know if you'd like me to add it back in a different spot.

Reply:
Yes that could be a problem. I think example tests may help.